File: sdks/swift/Sources/DoowTrack/DoowError.swift
import Foundation

public struct DoowError: Error, LocalizedError {
    public let message: String
    public let statusCode: Int
    public let errorClass: String?

    public init(_ message: String, statusCode: Int = 0, errorClass: String? = nil) {
        self.message = message
        self.statusCode = statusCode
        self.errorClass = errorClass
    }

    public var errorDescription: String? { message }

    public var isNotFound: Bool { statusCode == 404 }
    public var isUnauthorized: Bool { statusCode == 401 }
    public var isForbidden: Bool { statusCode == 403 }
    public var isRateLimited: Bool { statusCode == 429 }
    public var isServerError: Bool { statusCode >= 500 }
}

File: sdks/dotnet/src/DoowTrack/DoowError.cs
namespace DoowTrack;

public class DoowError : Exception
{
    public int StatusCode { get; }
    public string? ErrorClass { get; }

    public DoowError(string message, int statusCode = 0, string? errorClass = null)
        : base(message)
    {
        StatusCode = statusCode;
        ErrorClass = errorClass;
    }

    public bool IsNotFound() => StatusCode == 404;
    public bool IsUnauthorized() => StatusCode == 401;
    public bool IsForbidden() => StatusCode == 403;
    public bool IsRateLimited() => StatusCode == 429;
    public bool IsServerError() => StatusCode >= 500;
}
